File: packages/shared/src/schemas/news-item.ts
/**
 * NewsItem Schema
 * Represents a news article or update from various sources
 * 
 * @version 1.0.0
 * @backward-compatibility
 * - v1.0.0: Initial schema
 */

import { z } from "zod";
import {
  createVersionedSchema,
  sentimentSchema,
  severitySchema,
  CURRENT_SCHEMA_VERSION,
} from "./common.js";

// ============================================
// SCHEMA DEFINITION
// ============================================

export const newsSourceSchema = z.enum([
  "twitter",
  "discord",
  "telegram",
  "reddit",
  "medium",
  "mirror",
  "rss",
  "api",
  "scraper",
  "other",
]);

export type NewsSource = z.infer<typeof newsSourceSchema>;

export const newsCategorySchema = z.enum([
  "market",
  "protocol",
  "regulatory",
  "technical",
  "social",
  "security",
  "partnership",
  "listing",
  "other",
]);

export type NewsCategory = z.infer<typeof newsCategorySchema>;

export const newsItemSchema = createVersionedSchema({
  // Content
  title: z.string().min(1).max(500),
  content: z.string().min(1).max(50000),
  summary: z.string().max(1000).optional(),
  
  // Source information
  source: newsSourceSchema,
  sourceUrl: z.string().url().optional(),
  sourceId: z.string().optional(),
  author: z.string().optional(),
  
  // Classification
  category: newsCategorySchema,
  tags: z.array(z.string()).default([]),
  
  // Analysis
  sentiment: sentimentSchema.optional(),
  sentimentScore: z.number().min(-1).max(1).optional(),
  relevanceScore: z.number().min(0).max(1).optional(),
  importance: severitySchema.default("medium"),
  
  // Related entities
  mentionedTokens: z.array(z.string()).default([]),
  mentionedAddresses: z.array(z.string()).default([]),
  
  // Metadata
  language: z.string().default("en"),
  publishedAt: z.string().datetime(),
  fetchedAt: z.string().datetime(),
  
  // Processing status
  processed: z.boolean().default(false),
  embeddingId: z.string().uuid().optional(),
});

export type NewsItem = z.infer<typeof newsItemSchema>;

// ============================================
// FACTORY FUNCTIONS
// ============================================

export function createNewsItem(
  data: Omit<NewsItem, "id" | "schemaVersion" | "createdAt">
): NewsItem {
  return newsItemSchema.parse({
    ...data,
    id: crypto.randomUUID(),
    schemaVersion: CURRENT_SCHEMA_VERSION,
    createdAt: new Date().toISOString(),
  });
}

// ============================================
// EXAMPLES & FIXTURES
// ============================================

export const newsItemExamples: NewsItem[] = [
  {
    schemaVersion: "1.0.0",
    id: "550e8400-e29b-41d4-a716-446655440001",
    createdAt: "2024-01-15T10:30:00Z",
    title: "Monad Mainnet Launch Announcement",
    content: "The Monad team has announced the mainnet launch date for Q1 2024...",
    summary: "Monad mainnet launching Q1 2024",
    source: "twitter",
    sourceUrl: "https://twitter.com/moaborz/status/123456789",
    author: "moaborz",
    category: "protocol",
    tags: ["monad", "mainnet", "launch"],
    sentiment: "bullish",
    sentimentScore: 0.85,
    relevanceScore: 0.95,
    importance: "high",
    mentionedTokens: ["MON"],
    mentionedAddresses: [],
    language: "en",
    publishedAt: "2024-01-15T10:00:00Z",
    fetchedAt: "2024-01-15T10:30:00Z",
    processed: true,
    embeddingId: "550e8400-e29b-41d4-a716-446655440002",
  },
];

File: services/ingestion/src/schemas/news_item.rs
//! NewsItem Schema
//! 
//! Represents a news article or update from various sources
//! Compatible with TypeScript NewsItem schema

use serde::{Deserialize, Serialize};
use super::common::{Sentiment, Severity, Uuid, Timestamp, SchemaVersion};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum NewsSource {
    Twitter,
    Discord,
    Telegram,
    Reddit,
    Medium,
    Mirror,
    Rss,
    Api,
    Scraper,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum NewsCategory {
    Market,
    Protocol,
    Regulatory,
    Technical,
    Social,
    Security,
    Partnership,
    Listing,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewsItem {
    // Base fields
    pub schema_version: SchemaVersion,
    pub id: Uuid,
    pub created_at: Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<Timestamp>,
    
    // Content
    pub title: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    
    // Source information
    pub source: NewsSource,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<String>,
    
    // Classification
    pub category: NewsCategory,
    #[serde(default)]
    pub tags: Vec<String>,
    
    // Analysis
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sentiment: Option<Sentiment>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sentiment_score: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relevance_score: Option<f64>,
    #[serde(default = "default_importance")]
    pub importance: Severity,
    
    // Related entities
    #[serde(default)]
    pub mentioned_tokens: Vec<String>,
    #[serde(default)]
    pub mentioned_addresses: Vec<String>,
    
    // Metadata
    #[serde(default = "default_language")]
    pub language: String,
    pub published_at: Timestamp,
    pub fetched_at: Timestamp,
    
    // Processing status
    #[serde(default)]
    pub processed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding_id: Option<Uuid>,
}

fn default_importance() -> Severity {
    Severity::Medium
}

fn default_language() -> String {
    "en".to_string()
}

impl NewsItem {
    pub fn new(
        title: String,
        content: String,
        source: NewsSource,
        category: NewsCategory,
        published_at: Timestamp,
        fetched_at: Timestamp,
    ) -> Self {
        Self {
            schema_version: super::CURRENT_SCHEMA_VERSION.to_string(),
            id: uuid::Uuid::new_v4().to_string(),
            created_at: chrono::Utc::now().to_rfc3339(),
            updated_at: None,
            title,
            content,
            summary: None,
            source,
            source_url: None,
            source_id: None,
            author: None,
            category,
            tags: vec![],
            sentiment: None,
            sentiment_score: None,
            relevance_score: None,
            importance: Severity::Medium,
            mentioned_tokens: vec![],
            mentioned_addresses: vec![],
            language: "en".to_string(),
            published_at,
            fetched_at,
            processed: false,
            embedding_id: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_news_item_serialization() {
        let item = NewsItem::new(
            "Test Title".to_string(),
            "Test content".to_string(),
            NewsSource::Twitter,
            NewsCategory::Protocol,
            "2024-01-15T10:00:00Z".to_string(),
            "2024-01-15T10:30:00Z".to_string(),
        );

        let json = serde_json::to_string_pretty(&item).unwrap();
        println!("{}", json);
        
        // Verify camelCase
        assert!(json.contains("schemaVersion"));
        assert!(json.contains("createdAt"));
        assert!(json.contains("sourceUrl")); // Should not appear if None, but field name is camelCase
        
        // Deserialize back
        let parsed: NewsItem = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.title, item.title);
    }

    #[test]
    fn test_typescript_compatibility() {
        // JSON from TypeScript
        let ts_json = r#"{
            "schemaVersion": "1.0.0",
            "id": "550e8400-e29b-41d4-a716-446655440001",
            "createdAt": "2024-01-15T10:30:00Z",
            "title": "Monad Mainnet Launch",
            "content": "The Monad team has announced...",
            "source": "twitter",
            "category": "protocol",
            "tags": ["monad", "mainnet"],
            "sentiment": "bullish",
            "sentimentScore": 0.85,
            "importance": "high",
            "mentionedTokens": ["MON"],
            "mentionedAddresses": [],
            "language": "en",
            "publishedAt": "2024-01-15T10:00:00Z",
            "fetchedAt": "2024-01-15T10:30:00Z",
            "processed": true
        }"#;

        let parsed: NewsItem = serde_json::from_str(ts_json).unwrap();
        assert_eq!(parsed.title, "Monad Mainnet Launch");
        assert_eq!(parsed.sentiment, Some(Sentiment::Bullish));
        assert_eq!(parsed.importance, Severity::High);
    }
}
